examples/research_projects/codeparrot/scripts/human_eval_mgpu.py

```diff
@@ -0,0 +1,218 @@
+import json
+import multiprocessing
+import os
+import re
+
+from datasets import load_dataset, load_metric
+from tqdm import tqdm
+import torch
+from torch.utils.data import IterableDataset
+from torch.utils.data.dataloader import DataLoader
+
+from accelerate import Accelerator
+import transformers
+from arguments import HumanEvalArguments
+from transformers import (
+    AutoModelForCausalLM,
+    AutoTokenizer,
+    HfArgumentParser,
+    StoppingCriteria,
+    StoppingCriteriaList,
+    set_seed,
+)
+
+
+EOF_STRINGS = ["\nclass", "\ndef", "\n#", "\n@", "\nprint", "\nif"]
+
+
+class TokenizeDataset(IterableDataset):
+    """Tokenize and preprocess the dataset
+    Multiple copies of the same prompt are sent sequentially.
+    See compute_code for more details.
+    """
+    def __init__(self, tokenizer, dataset, n_tasks=None, n_copies=1):
+        self.tokenizer = tokenizer
+        self.dataset = dataset
+        self.n_tasks = len(dataset) if n_tasks is None else n_tasks
+        self.n_copies = n_copies
+
+    def __iter__(self):
+        for task in range(self.n_tasks):
+            # without strip, the model generate commented codes ...
+            prompt = self.tokenizer.eos_token + self.dataset[task]['prompt'].strip()
+            # codeparrot model is not robust to padding
+            input_ids = self.tokenizer(prompt)['input_ids']
+            for _ in range(self.n_copies):
+                yield {"ids": torch.tensor(input_ids)}
```

Comment:
If we move to the more general way to distribute the task with the padding, it might become harder to track which sample belongs to which task. We could also yield the `task_id` so that after the generation we can easily match all the generations with the tasks again.